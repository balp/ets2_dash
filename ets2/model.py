import json
from dataclasses import dataclass
from typing import Optional, Dict, List

from ets2.config import TruckConfig, truckconfig_from_dict, TrailerConfig, trailer_config_from_dict
from ets2.jobs import JobConfig, jobconfig_from_dict
from ets2.telematic import Telematic, telematic_from_dict
from ets2.tracks import Tracks


@dataclass
class Info:
    paused: bool


def info_from_dict(data: Dict) -> Info:
    return Info(paused=data['paused'])


@dataclass
class Version:
    major: int
    minor: int


def version_from_dict(data: Dict) -> Version:
    return Version(data['major'], data['minor'])


@dataclass
class Game:
    id: str
    name: str
    raw_version: int
    version: Version


def game_from_dict(data: Dict) -> Game:
    return Game(id=data['id'],
                name=data['name'],
                raw_version=data['raw_version'],
                version=version_from_dict(data['version']))


class Model:
    def __init__(self):
        self._observers = []
        self.telematic: Optional[Telematic] = None
        self.job: Optional[JobConfig] = None
        self.info: Optional[Info] = None
        self.game: Optional[Game] = None
        self.truck_config: Optional[TruckConfig] = None
        self.trailer_config: List[Optional[TrailerConfig]] = [None for _ in range(10)]
        self.tracks: Tracks = Tracks()

    def register_observer(self, observer):
        self._observers.append(observer)

    def _data_changed(self, change):
        for observer in self._observers:
            observer.notify(self, change)

    def set_telematic_data(self, data):
        self.telematic = telematic_from_dict(data)
        self.tracks.add_telematic(self.telematic)
        self._data_changed("telematic")

    def set_job_config(self, data):
        self.job = jobconfig_from_dict(data)
        self._data_changed("job")

    def set_info(self, data):
        self.info = info_from_dict(data)
        self._data_changed("info")

    def set_game(self, data):
        self.game = game_from_dict(data)
        self._data_changed("game")

    def set_truck_config(self, data):
        self.truck_config = truckconfig_from_dict(data)
        self._data_changed("truck_config")

    def set_trailer_config(self, data, index):
        self.trailer_config[index] = trailer_config_from_dict(data)
        self._data_changed("trailer_config")

    def get_game_pause(self) -> Optional[bool]:
        if self.info:
            return self.info.paused
        return None

    def get_game_name(self) -> Optional[str]:
        if self.game:
            return self.game.name
        return None

    def get_game_id(self) -> Optional[str]:
        if self.game:
            return self.game.id
        return None

    def get_time_left(self) -> Optional[int]:
        if self.telematic and self.job:
            game_time = self.telematic.common.game_time
            delivery_time: int = self.job.delivery_time
            if game_time and delivery_time and delivery_time != 0xFFFFFFFF:
                return delivery_time - game_time
        return None

    def get_time_to_rest(self) -> int:
        if self.telematic:
            return self.telematic.common.rest_stop
        return 0

    def get_time_destination(self) -> int:
        if self.telematic:
            eta = self.telematic.truck.navigation_time
            return int(eta) // 60
        return 0

    def get_time_destination_with_rest(self) -> int:
        if self.telematic:

            drive_time = self.get_time_destination()
            next_rest = self.get_time_to_rest()
            total_time = drive_time

            if drive_time > next_rest:
                total_time += 9*60
                total_time += ((drive_time - next_rest) // (11 * 60)) * 9 * 60
            return int(total_time)
        return 0

    def get_speed_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 3.6
        return 0.0

    def get_cruise_control_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 3.6
        return 0.0

    def get_speed_limit_kmh(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 3.6
        return 0.0

    def get_speed_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.speed * 2.2369363
        return 0.0

    def get_cruise_control_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.cruise_control * 2.2369363
        return 0.0

    def get_speed_limit_mph(self) -> float:
        if self.telematic:
            return self.telematic.truck.navigation_speed_limit * 2.2369363
        return 0.0

    def get_fuel_left(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_amount
        return 0.0

    def get_fuel_range(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_range
        return 0.0

    def get_fuel_consumtion(self) -> float:
        if self.telematic:
            return self.telematic.truck.fuel_consumption_average
        return 0.0

    def get_wear_cabin(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_cabin
        return 0.0

    def get_wear_chassis(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_chassis
        return 0.0

    def get_wear_engine(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_engine
        return 0.0

    def get_wear_transmission(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_transmission
        return 0.0

    def get_wear_wheels(self) -> float:
        if self.telematic:
            return self.telematic.truck.wear_wheels
        return 0.0

    def get_wear_trailer(self) -> float:
        if self.telematic:
            return self.telematic.trailer.wear_chassis
        return 0.0

    def get_light_high_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_high
        return False

    def get_light_low_beam(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beam_low
        return False

    def get_light_beacon(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_beacon
        return False

    def get_light_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_lblinker
        return False

    def get_light_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_rblinker
        return False

    def get_l_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.lblinker
        return False

    def get_r_blinker(self) -> bool:
        if self.telematic:
            return self.telematic.truck.rblinker
        return False

    def get_light_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_parking
        return False

    def get_light_reverse(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_reverse
        return False

    def get_light_aux_front(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_front != 0
        return False

    def get_light_aux_roof(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_aux_roof != 0
        return False

    def get_light_breaking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.light_brake
        return False

    def get_ad_blue_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.adblue_warning
        return False

    def get_break_emergency(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_emergency
        return False

    def get_break_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure_warning
        return False

    def get_break_motor(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_motor
        return False

    def get_break_parking(self) -> bool:
        if self.telematic:
            return self.telematic.truck.brake_parking
        return False

    def get_electric(self) -> bool:
        if self.telematic:
            return self.telematic.truck.electric_enabled
        return False

    def get_battery_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.battery_voltage_warning
        return False

    def get_engine(self) -> bool:
        if self.telematic:
            return self.telematic.truck.engine_enabled
        return False

    def get_fuel_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.fuel_warning
        return False

    def get_oil_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.oil_pressure_warning
        return False

    def get_water_warning(self) -> bool:
        if self.telematic:
            return self.telematic.truck.water_temperature_warning
        return False

    def get_wipers(self) -> bool:
        if self.telematic:
            return self.telematic.truck.wipers
        return False

    def get_air_pressure(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_air_pressure
        return False

    def get_break_retarder(self) -> int:
        if self.telematic:
            return self.telematic.truck.brake_retarder
        return False

    def get_break_temperature(self) -> float:
        if self.telematic:
            return self.telematic.truck.brake_temperature
        return False


def add_json_to_model(model: Model, json_data: json, topic: str):
    if topic == "ets2/data":
        model.set_telematic_data(json_data)
    elif topic == "ets2/game":
        model.set_game(json_data)
    elif topic == "ets2/info":
        model.set_info(json_data)
    elif topic == "ets2/info/config/job":
        model.set_job_config(json_data)
    elif topic == "ets2/info/config/truck":
        model.set_truck_config(json_data)
    elif topic == "ets2/info/config/trailer.0":
        model.set_trailer_config(json_data, 0)
    elif topic == "ets2/info/config/trailer.1":
        model.set_trailer_config(json_data, 1)
    elif topic == "ets2/info/config/trailer.2":
        model.set_trailer_config(json_data, 2)
    elif topic == "ets2/info/config/trailer.3":
        model.set_trailer_config(json_data, 3)
    elif topic == "ets2/info/config/trailer.4":
        model.set_trailer_config(json_data, 4)
    elif topic == "ets2/info/config/trailer.5":
        model.set_trailer_config(json_data, 5)
    elif topic == "ets2/info/config/trailer.6":
        model.set_trailer_config(json_data, 6)
    elif topic == "ets2/info/config/trailer.7":
        model.set_trailer_config(json_data, 7)
    elif topic == "ets2/info/config/trailer.8":
        model.set_trailer_config(json_data, 8)
    elif topic == "ets2/info/config/trailer.9":
        model.set_trailer_config(json_data, 9)
