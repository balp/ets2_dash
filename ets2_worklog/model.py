import datetime
import json
from dataclasses import dataclass
from typing import Optional, Dict, List

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[datetime.datetime]
    ended: Optional[datetime.datetime]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model):
        self._model = data
        self._model.register_observer(self)
        self._jobs: List[Job] = []

    def __repr__(self):
        return str(self._jobs)

    def notify(self, model: ets2.model.Model, change: str):
        if self._jobs:
            if model.job != self._jobs[-1].config:
                self._jobs[-1].ended = self.time_in_game(model)
                job = self.job_from_model(model)
                self._jobs.append(job)
            else:
                if model.telematic:
                    self._jobs[-1].track.add_telematic(model.telematic)
        else:
            job = self.job_from_model(model)
            self._jobs.append(job)

    def job_from_model(self, model):
        started = self.time_in_game(model)
        job = Job(config=model.job,
                  started=started,
                  ended=None,
                  delivered=None,
                  cancelled=None,
                  track=ets2.model.Tracks())
        return job

    def time_in_game(self, model):
        if model.telematic:
            started = datetime.datetime.fromordinal(1) \
                      + datetime.timedelta(minutes=model.telematic.common.game_time)
        else:
            started = None
        return started

    def job_delivered(self, delivered: Delivered) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
