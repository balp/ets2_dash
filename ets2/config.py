#  Copyright (c) 2020. Anders Arnholm <Anders@Arnholm.se>
#
#   Permission to use, copy, modify, and distribute this software for any
#   purpose with or without fee is hereby granted, provided that the above
#   copyright notice and this permission notice appear in all copies.
#
#   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
#   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
#   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
#   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
#   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
#   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
#   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
from dataclasses import dataclass
from typing import Dict, Optional

from ets2.types import Vector, vector_from_dict


@dataclass
class TruckConfig:
    id: str
    name: str
    adblue_capacity: float
    adblue_warning_factor: float
    battery_voltage_warning: float
    brake_air_pressure_emergency: float
    brake_air_pressure_warning: float
    brand: str
    brand_id: str
    cabin_position: Vector
    differential_ratio: float
    forward_ratio: float
    fuel_capacity: float
    fuel_warning_factor: float
    gears_forward: int
    gears_reverse: int
    head_position: Vector
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    oil_pressure_warning: float
    retarder_steps: int
    reverse_ratio: float
    rpm_limit: float
    water_temperature_warning: float
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: bool


def truckconfig_from_dict(data: Dict) -> Optional[TruckConfig]:
    if len(data):
        return TruckConfig(
            id=data['id'],
            name=data['name'],
            adblue_capacity=data['adblue.capacity'],
            adblue_warning_factor=data['adblue.warning.factor'],
            battery_voltage_warning=data['battery.voltage.warning'],
            brake_air_pressure_emergency=data['brake.air.pressure.emergency'],
            brake_air_pressure_warning=data['brake.air.pressure.warning'],
            brand=data['brand'],
            brand_id=data['brand_id'],
            cabin_position=vector_from_dict(data['cabin.position']),
            differential_ratio=data['differential.ratio'],
            forward_ratio=data['forward.ratio'],
            fuel_capacity=data['fuel.capacity'],
            fuel_warning_factor=data['fuel.warning.factor'],
            gears_forward=data['gears.forward'],
            gears_reverse=data['gears.reverse'],
            head_position=vector_from_dict(data['head.position']),
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            oil_pressure_warning=data['oil.pressure.warning'],
            retarder_steps=data['retarder.steps'],
            reverse_ratio=data['reverse.ratio'],
            rpm_limit=data['rpm.limit'],
            water_temperature_warning=data['water.temperature.warning'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None


@dataclass
class TrailerConfig:
    id: str
    body_type: Optional[str]
    chain_type: Optional[str]
    cargo_accessory_id: str
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: int


def trailer_config_from_dict(data: Dict) -> Optional[TrailerConfig]:
    if len(data):
        return TrailerConfig(
            id=data['id'],
            body_type=(data['body.type'] if 'body.type' in data else None),
            chain_type=(data['chain.type'] if 'chain.type' in data else None),
            cargo_accessory_id=data['cargo.accessory.id'],
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None