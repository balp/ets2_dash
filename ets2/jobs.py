from dataclasses import dataclass
from typing import Optional, Dict

import ets2.model
import ets2.tracks


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class JobConfig:
    cargo: str
    cargo_id: str
    cargo_mass: float
    delivery_time: int
    destination_city: str
    destination_city_id: str
    destination_company: str
    destination_company_id: str
    income: int
    source_city: str
    source_city_id: str
    source_company: str
    source_company_id: str


def jobconfig_from_dict(data: Dict) -> Optional[JobConfig]:
    if len(data):
        return JobConfig(cargo=data['cargo'],
                         cargo_id=data['cargo.id'],
                         cargo_mass=data['cargo.mass'],
                         delivery_time=data['delivery.time'],
                         destination_city=data['destination.city'],
                         destination_city_id=data['destination.city.id'],
                         destination_company=data['destination.company'],
                         destination_company_id=data['destination.company.id'],
                         income=data['income'],
                         source_city=data['source.city'],
                         source_city_id=data['source.city.id'],
                         source_company=data['source.company'],
                         source_company_id=data['source.company.id'])
    else:
        return None


@dataclass
class Job:
    config: Optional[JobConfig]
    started: Optional[int]
    ended: Optional[int]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.tracks.Tracks
    id: Optional[int]

