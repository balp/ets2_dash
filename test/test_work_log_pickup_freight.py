import bz2
import json

import jsonpickle as jsonpickle
import pytest
from approvaltests import verify_with_namer, get_default_reporter
from approvaltests.core.namer import Namer

from ets2.model import Model, add_json_to_model
from ets2_worklog.model import WorkLog

def test_pickup_freight():
    model = Model()
    work_log: WorkLog = WorkLog(model)

    with bz2.open("data/ats_start_port_ang_coos_bay.mqtt.json.bz2") as gj:
        for line in gj:
            topic, json_str = line.decode('utf8').split(" ", 1)
            json_data = json.loads(json_str)
            add_json_to_model(model, json_data, topic)
    jsonpickle.set_encoder_options('json', sort_keys=True, indent=4)
    jsonpickle.set_preferred_backend('json')
    verify_with_namer(data=jsonpickle.encode(work_log),
                      namer=Namer(extension='.json'),
                      reporter=get_default_reporter())

def test_deliver_freight():
    model = Model()
    work_log: WorkLog = WorkLog(model)
    for f in ["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
              "data/ats_end_port_ang_coos_bay.mqtt.json.bz2"]:
        with bz2.open(f) as gj:
            for line in gj:
                topic, json_str = line.decode('utf8').split(" ", 1)
                json_data = json.loads(json_str)
                add_json_to_model(model, json_data, topic)
        jsonpickle.set_encoder_options('json', sort_keys=True, indent=4)
    jsonpickle.set_preferred_backend('json')
    verify_with_namer(data=jsonpickle.encode(work_log),
                      namer=Namer(extension='.json'),
                      reporter=get_default_reporter())



