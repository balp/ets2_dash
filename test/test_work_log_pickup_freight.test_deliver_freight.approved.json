{
    "_db": {
        "_conn": {
            "py/object": "sqlite3.Connection"
        },
        "py/object": "ets2.database.DataBase"
    },
    "_model": {
        "_observers": [
            {
                "py/id": 0
            }
        ],
        "game": {
            "id": "ats",
            "name": "American Truck Simulator 1.36.1.34s",
            "py/object": "ets2.model.Game",
            "raw_version": 65539,
            "version": {
                "major": 1,
                "minor": 3,
                "py/object": "ets2.model.Version"
            }
        },
        "info": {
            "paused": false,
            "py/object": "ets2.model.Info"
        },
        "job": null,
        "py/object": "ets2.model.Model",
        "telematic": {
            "common": {
                "game_time": 38312,
                "py/object": "ets2.telematic.Common",
                "rest_stop": 581,
                "scale": 3.0
            },
            "py/object": "ets2.telematic.Telematic",
            "trailer": {
                "acceleration_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": -0.0011283610947430134,
                    "y": 1.2916611922264565e-06,
                    "z": 5.581895948125748e-06
                },
                "acceleration_linear": {
                    "py/object": "ets2.types.Vector",
                    "x": 2.687186497496441e-05,
                    "y": -0.013560137711465359,
                    "z": 0.0021443134173750877
                },
                "cargo_damage": 6.945314817130566e-05,
                "connected": false,
                "py/object": "ets2.telematic.Trailer",
                "velocity_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": 0.00011189553333679214,
                    "y": -2.2064783422592882e-07,
                    "z": 4.46151375399495e-07
                },
                "velocity_linear": {
                    "py/object": "ets2.types.Vector",
                    "x": 8.353917110071052e-06,
                    "y": 0.0024038839619606733,
                    "z": -0.00010401262989034876
                },
                "wear_chassis": 0.013595662079751492,
                "wear_wheels": 0.013595662079751492,
                "world_placement": {
                    "orientation": {
                        "heading": 0.3958577811717987,
                        "pitch": 8.363186498172581e-05,
                        "py/object": "ets2.types.Euler",
                        "roll": -1.0700628990889527e-05
                    },
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -113930.79376220703,
                        "y": 5.173255920410156,
                        "z": -42782.45690917969
                    },
                    "py/object": "ets2.types.Placement"
                }
            },
            "truck": {
                "adblue": 64.88665771484375,
                "adblue_warning": false,
                "battery_voltage": 13.69074821472168,
                "battery_voltage_warning": false,
                "brake_air_pressure": 97.89523315429688,
                "brake_air_pressure_emergency": false,
                "brake_air_pressure_warning": false,
                "brake_motor": false,
                "brake_parking": false,
                "brake_retarder": 0,
                "brake_temperature": 94.8542251586914,
                "cabin_acceleration_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cabin_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": 0.0,
                        "py/object": "ets2.types.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": 0.0,
                        "y": 0.0,
                        "z": 0.0
                    },
                    "py/object": "ets2.types.Placement"
                },
                "cabin_velocity_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cruise_control": 0.0,
                "dashboard_backlight": 1.0,
                "displayed_gear": 3,
                "effective_brake": 1.0,
                "effective_clutch": 0.0,
                "effective_steering": 0.002087546046823263,
                "effective_throttle": 0.0,
                "electric_enabled": true,
                "engine_enabled": true,
                "engine_gear": 0,
                "engine_rpm": 599.9986572265625,
                "fuel_amount": 455.7331848144531,
                "fuel_consumption_average": 0.46223539113998413,
                "fuel_range": 1431.0462646484375,
                "fuel_warning": false,
                "head_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": -0.01944444701075554,
                        "py/object": "ets2.types.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": 0.0,
                        "y": -0.019999999552965164,
                        "z": 0.0
                    },
                    "py/object": "ets2.types.Placement"
                },
                "hshifter_slot": 0,
                "input_brake": 0.0,
                "input_clutch": 0.0,
                "input_steering": 0.002087546046823263,
                "input_throttle": 0.0,
                "lblinker": false,
                "light_aux_front": 1,
                "light_aux_roof": 1,
                "light_beacon": false,
                "light_beam_high": false,
                "light_beam_low": true,
                "light_brake": false,
                "light_lblinker": false,
                "light_parking": true,
                "light_rblinker": false,
                "light_reverse": false,
                "local_acceleration_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": -0.003223245730623603,
                    "y": 5.16258387506241e-06,
                    "z": 0.00010118159116245806
                },
                "local_acceleration_linear": {
                    "py/object": "ets2.types.Vector",
                    "x": 0.00037163091474212706,
                    "y": -0.0057870023883879185,
                    "z": 0.023027215152978897
                },
                "local_velocity_angular": {
                    "py/object": "ets2.types.Vector",
                    "x": -0.0004103201499674469,
                    "y": 3.180055045959307e-06,
                    "z": 1.4616225598729216e-05
                },
                "local_velocity_linear": {
                    "py/object": "ets2.types.Vector",
                    "x": 8.843670366331935e-05,
                    "y": -0.0013863801723346114,
                    "z": 0.0057382117956876755
                },
                "navigation_distance": 737.828857421875,
                "navigation_speed_limit": 0.0,
                "navigation_time": 117.77947998046875,
                "odometer": 17137.17578125,
                "oil_pressure": 39.54541015625,
                "oil_pressure_warning": false,
                "oil_temperature": 90.75440979003906,
                "py/object": "ets2.telematic.Truck",
                "rblinker": false,
                "speed": -0.00712883984670043,
                "water_temperature": 86.2168197631836,
                "water_temperature_warning": false,
                "wear_cabin": 0.00254922010935843,
                "wear_chassis": 0.0031865250784903765,
                "wear_engine": 0.0037152734585106373,
                "wear_transmission": 0.0030779684893786907,
                "wear_wheels": 0.021697821095585823,
                "wipers": false,
                "world_placement": {
                    "orientation": {
                        "heading": 0.3937174081802368,
                        "pitch": 4.819077730644494e-05,
                        "py/object": "ets2.types.Euler",
                        "roll": -1.89727472843515e-07
                    },
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -113934.71926879883,
                        "y": 5.177162170410156,
                        "z": -42777.36328125
                    },
                    "py/object": "ets2.types.Placement"
                }
            }
        },
        "tracks": {
            "last_time": 38312,
            "points": [
                {
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -104305.18292236328,
                        "y": 12.102846145629883,
                        "z": -66655.20185089111
                    },
                    "py/object": "ets2.tracks.TrackPoint",
                    "time": 37201
                },
                {
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -104272.02157592773,
                        "y": 12.105484962463379,
                        "z": -66604.86669540405
                    },
                    "py/object": "ets2.tracks.TrackPoint",
                    "time": 37228
                },
                {
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -113779.6833114624,
                        "y": 5.177712440490723,
                        "z": -42559.202308654785
                    },
                    "py/object": "ets2.tracks.TrackPoint",
                    "time": 38299
                },
                {
                    "position": {
                        "py/object": "ets2.types.Vector",
                        "x": -113934.7191772461,
                        "y": 5.177143096923828,
                        "z": -42777.36331176758
                    },
                    "py/object": "ets2.tracks.TrackPoint",
                    "time": 38312
                }
            ],
            "py/object": "ets2.tracks.Tracks"
        },
        "trailer_config": [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null
        ],
        "truck_config": {
            "adblue_capacity": 80.0,
            "adblue_warning_factor": 0.15000000596046448,
            "battery_voltage_warning": 11.880000114440918,
            "brake_air_pressure_emergency": 29.579999923706055,
            "brake_air_pressure_warning": 59.15999984741211,
            "brand": "Volvo",
            "brand_id": "volvo",
            "cabin_position": {
                "py/object": "ets2.types.Vector",
                "x": 0.0,
                "y": 3.0,
                "z": -2.0
            },
            "differential_ratio": 3.549999952316284,
            "forward_ratio": 0.7300000190734863,
            "fuel_capacity": 758.0,
            "fuel_warning_factor": 0.15000000596046448,
            "gears_forward": 13,
            "gears_reverse": 3,
            "head_position": {
                "py/object": "ets2.types.Vector",
                "x": -0.5642366409301758,
                "y": -0.5814938545227051,
                "z": 0.570602536201477
            },
            "hook_position": {
                "py/object": "ets2.types.Vector",
                "x": 0.0,
                "y": 1.0,
                "z": 1.068235158920288
            },
            "id": "vehicle.volvo.vnl",
            "license_plate": "8A7P001",
            "license_plate_country": "California",
            "license_plate_country_id": "california",
            "name": "VNL",
            "oil_pressure_warning": 10.149999618530273,
            "py/object": "ets2.config.TruckConfig",
            "retarder_steps": 3,
            "reverse_ratio": -4.03000020980835,
            "rpm_limit": 2100.0,
            "water_temperature_warning": 105.0,
            "wheel_position": {
                "py/object": "ets2.types.Vector",
                "x": 0.9612910747528076,
                "y": 0.5059999823570251,
                "z": 0.5031635761260986
            },
            "wheel_powered": true,
            "wheel_radius": 0.5072842240333557,
            "wheel_simulated": true,
            "wheel_steerable": false,
            "wheels_count": 6
        }
    },
    "jobs": [
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": 37228,
            "id": 1,
            "py/object": "ets2.jobs.Job",
            "started": 37201,
            "track": {
                "last_time": 37228,
                "points": [
                    {
                        "position": {
                            "py/object": "ets2.types.Vector",
                            "x": -104305.18292236328,
                            "y": 12.102846145629883,
                            "z": -66655.20185089111
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 37201
                    },
                    {
                        "position": {
                            "py/id": 39
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 37228
                    }
                ],
                "py/object": "ets2.tracks.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": {
                "cargo": "Lumber",
                "cargo_id": "lumber_b",
                "cargo_mass": 21387.240234375,
                "delivery_time": 38821,
                "destination_city": "Coos Bay",
                "destination_city_id": "coos_bay",
                "destination_company": "Olthon Homes",
                "destination_company_id": "oh_con_hom",
                "income": 18742,
                "py/object": "ets2.jobs.JobConfig",
                "source_city": "Port Angeles",
                "source_city_id": "port_angeles",
                "source_company": "Deepgrove",
                "source_company_id": "dg_wd_saw"
            },
            "delivered": {
                "auto_load": true,
                "auto_park": false,
                "cargo_damage": 0.0,
                "distance": 676.0,
                "py/object": "ets2.jobs.Delivered",
                "revenue": 18742,
                "time": 1084,
                "xp": 870
            },
            "ended": 38312,
            "id": 2,
            "py/object": "ets2.jobs.Job",
            "started": 37228,
            "track": {
                "last_time": 38312,
                "points": [
                    {
                        "position": {
                            "py/id": 39
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 37228
                    },
                    {
                        "position": {
                            "py/id": 41
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 38299
                    },
                    {
                        "position": {
                            "py/id": 43
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 38312
                    }
                ],
                "py/object": "ets2.tracks.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": null,
            "id": 3,
            "py/object": "ets2.jobs.Job",
            "started": 38312,
            "track": {
                "last_time": 38312,
                "points": [
                    {
                        "position": {
                            "py/object": "ets2.types.Vector",
                            "x": -113934.7191772461,
                            "y": 5.177135944366455,
                            "z": -42777.36331176758
                        },
                        "py/object": "ets2.tracks.TrackPoint",
                        "time": 38312
                    }
                ],
                "py/object": "ets2.tracks.Tracks"
            }
        }
    ],
    "py/object": "ets2.work_log.WorkLog"
}